### PR TITLE
executer: make updateRecord easier to understand

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
`updateRecord` in write.go is too complex to understand. This pr tries to make it more clearly.

### What is changed and how it works?
1. move the null check to before the auto ID rebase, which can avoid an extra null check for auto ID.
1. use GetInt64() instead of ToInt64() because the kind of datum is already int when rebasing auto ID.
1. merge the check of `BadNullAsWarning` with null check together.
1. add some comments with steps.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

PTAL @coocood

### CODE REVIEW

File: executor/write.go [hunk 1]
```diff
@@ -44,114 +44,118 @@ var (
...
-			if errTI != nil {
-				return false, handleChanged, newHandle, errors.Trace(errTI)
-			}
-			err := t.RebaseAutoID(ctx, val, true)
-			if err != nil {
-				return false, handleChanged, newHandle, errors.Trace(err)
+	// 2. Check null.
+	for i, col := range t.Cols() {
+		if err := col.CheckNotNull(newData[i]); err != nil {
+			if sc.BadNullAsWarning {
```
Comment (coocood) [inline executor/write.go]: The old code doesn't have this logic?

File: executor/write.go [hunk 2]
```diff
@@ -44,114 +44,118 @@ var (
...
 			err = tables.CheckHandleExists(ctx, t, newHandle, newData)
 			if err != nil {
 				return false, handleChanged, newHandle, errors.Trace(err)
 			}
 			skipHandleCheck = true
 		}
-		err = t.RemoveRecord(ctx, h, oldData)
-		if err != nil {
-			return false, handleChanged, newHandle, errors.Trace(err)
+		if err := t.RemoveRecord(ctx, h, oldData); err != nil {
```
Comment (coocood) [inline executor/write.go]: This error shadows the outer one?

File: executor/write.go (hunk 1 above)
Comment (jackysp) [reply executor/write.go]: It has this at old file line 67, 76 and 109.
Line 67 is used for BadNullAsWarning only, line 76 is used for auto ID and line 109 checks all the not null constraint.

File: executor/write.go (hunk 2 above)
Comment (jackysp) [reply executor/write.go]: Oh, I'll check it as soon as I fix the `make check` on my laptop.

**Comment by jackysp** [comment]
PTAL @coocood

File: executor/write.go [hunk 3]
```diff
@@ -44,114 +44,117 @@ var (
...
-			modified[i] = true
-			// Rebase auto increment id if the field is changed.
-			if mysql.HasAutoIncrementFlag(col.Flag) {
-				if newData[i].IsNull() {
-					return false, handleChanged, newHandle, table.ErrColumnCantNull.GenByArgs(col.Name)
+	}
+
+	// 3. Compare datum, then handle some flags.
+	for i, col := range t.Cols() {
+		if cmp, err := newData[i].CompareDatum(sc, &oldData[i]); err == nil {
```
Comment (zz-jason) [inline executor/write.go]: how about:
```go
cmp, err := newData[i].CompareDatum(sc, &oldData[i])
if err != nil {
	return false, false, 0, errors.Trace(err)
}
if cmp != 0 {
	...
} else {
	...
}
```

**Comment by jackysp** [comment]
PTAL @zz-jason

**Comment by jackysp** [comment]
PTAL @zz-jason

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**